python/tskit/combinatorics.py

```diff
@@ -28,12 +28,133 @@
 import functools
 import heapq
 import itertools
+import json
+import random
 
+import attr
 import numpy as np
 
 import tskit
 
 
+@attr.s(eq=False)
+class TreeNode:
+    """
+    Simple linked tree class used to generate tree topologies.
+    """
+
+    parent = attr.ib(default=None)
+    children = attr.ib(factory=list)
+    label = attr.ib(default=None)
+
+    @staticmethod
+    def random_binary_tree(leaf_labels, rng):
+        """
+        Based on the description of Remy's method of generating "decorated"
+        random binary trees in TAOCP 7.2.1.6. This is not a direct
+        implementation of Algorithm R, because we are interested in
+        the leaf node labellings.
+
+        The pre-fasicle text is available here, page 16:
+        http://www.cs.utsa.edu/~wagner/knuth/fasc4a.pdf
+        """
+        nodes = [TreeNode(label=leaf_labels[0])]
+        for label in leaf_labels[1:]:
+            n = TreeNode(label=label)
+            x = rng.choice(nodes)
+            internal = TreeNode(parent=x.parent, children=[x, n])
+            if x.parent is not None:
+                index = x.parent.children.index(x)
+                x.parent.children[index] = internal
+            rng.shuffle(internal.children)
```

Comment:
I must admit that I don't quire understand why this extra shuffle is in there? Won't the first `rng.choice` make the selection random anyway?

Reply:
(by the way, I think this is essentially the same edge-addition algorithm that I originally implemented)

Reply:
We choose a node x to modify randomly in the first step, adding a new internal node which will be the parent of x and the n'th leaf. We must randomise the order of [x, n] because the leaf would aways be either the left or right child, which would (I think) mean we don't generate all possible trees.

I'll add a few comments to clarify, as it's a simple algorithm and worth explaining for future reference.

Reply:
I think it *does* generate all possible trees, even if the child order isn't shuffled. At least, it still did in my tests. And from reading the Knuth explanation, it doesn't seem necessary (I think?)